src/basic/copy.c

```diff
@@ -386,14 +386,13 @@ static int fd_copy_directory(
                         continue;
                 }
 
-                if (buf.st_dev != original_device)
-                        continue;
-
                 if (S_ISREG(buf.st_mode))
                         q = fd_copy_regular(dirfd(d), de->d_name, &buf, fdt, de->d_name, override_uid, override_gid, copy_flags);
-                else if (S_ISDIR(buf.st_mode))
+                else if (S_ISDIR(buf.st_mode)) {
+                        if (buf.st_dev != original_device)
+                                continue;
```

Comment:
The stat call on device nodes does not report the device number of the device itself but of the filesystem hosting the device node, so this one is fine. You are right though that this changes the behaviour for bind-mounted files which would be copied with my patch while they are skipped currently. That said copy_tree is only used by `systemd-tmpfiles` to setup the temporary files and by `machined` to copy files between host and containers. In both cases, I don't believe that we have any specific requirement to ignore bind mounted files.